### ISSUE TITLE
BUS not working correctly

### ISSUE THREAD

**Issue by thrust26** [stella-emu/stella#707]
Some old BUS demos from SpiceWare are not working as described and on real hardware. 

https://atariage.com/forums/topic/258191-bus-stuffing-demos/?do=findComment&comment=3676378

[128bus_20170120.zip](https://github.com/stella-emu/stella/files/5294285/128bus_20170120.zip)

**Comment by sa666666** [comment]
Yep, someone was discussing making use of versioning, so old BUS ROMs would work.  It would be really nice to get these working again, since I look on this as a regression (they _did_ work in an older version of Stella).  I suspect the best approach is to find the last time they worked in Stella, and then see what code changed.  Some stuff in `CartBUS` obviously changed, but I think some was in `Thumbulator` too.

**Comment by thrust26** [comment]
Darrell plans to update the demos to the later BUS driver, but maybe we are faster. 😈

**Comment by sa666666** [comment]
This is the commit that broke the ROMS: https://github.com/stella-emu/stella/commit/490724c07978317e74c6491d5084fa58056fd805.  Don't know how feasible it is to add support; perhaps we should wait for @SpiceWare to update the ROMs??

**Comment by thrust26** [comment]
Yes, probably better. The changes are too large and we do not really understand how the code works

**Comment by SpiceWare** [comment]
I seem to recall @sa666666 had decided we wouldn't support the preliminary BUS driver since only some tech demos were done with it. That was a few years ago, so I could have misremembered it.

@thrust26 - if there's a specific demo you're interested in I could take a look at it this week.

**Comment by thrust26** [comment]
@SpiceWare The one posted above would be nice.

**Comment by SpiceWare** [comment]
Will do.

**Comment by sa666666** [comment]
No, that's probably what I said.  While it would be nice to have the ROMs working, it seems like a lot of work to change the scheme and never have it used again.  Probably better to spend the time on the ROMs themselves.  Of course, only you @SpiceWare can work on that part, so it all falls to what you want to do.

**Comment by thrust26** [comment]
Yes, changing the ROMs, that's what I meant too.

**Comment by SpiceWare** [comment]
Forgot how primitive the old build process was - dasm would echo out #define statements that needed to be manually copied/pasted into a c header file.

Decided it would be best to migrate it to the new build process, seen in the CDFJ tutorial, that uses a makefile to do everything.  The c header file is now automatically created by parsing the dasm symbol file for all symbols beginning with an underscore character.

Took a while, but finally got the first image to show.  Over the next week or two I'll migrate the rest of the 128 pixel demo routines over, clean up the code, then post it in the BUS subforum of the Harmony/Melody club.

<img width="640" alt="Screen Shot 2020-10-25 at 7 35 51 PM" src="https://user-images.githubusercontent.com/26531922/97123361-7b865100-16f9-11eb-841e-f26c4f000ddf.png">

**Comment by sa666666** [comment]
Looking back at this again, it seems that Stella is freezing on those old BUS ROMs.  It would be nice if we could detect those old ROMs, and just display a message saying they're not supported.

**Comment by sa666666** [comment]
@SpiceWare, has this been resolved, or will you have time to work on this soon?  Just attempting to clean up the Stella Github issues list.

**Comment by SpiceWare** [comment]
Sorry, RL has kept me away from Atari projects for so long I'd forgotten about this and #745 

Looks like I'd started porting the BUS Demos to the last version of the driver.  I'll try to resume work on that this weekend. I'll also look into detecting if a ROM is using an obsolete BUS driver so an unsupported message can be shown.

**Comment by thrust26** [comment]
The PlusROM initialization was missing in `CartBUS.cxx`. I committed a fix, please try again.

**Comment by SpiceWare** [comment]
Thanks! Made some good headway this evening.  So far I've identified 3 versions of the BUS driver in the demos that are linked to in the first post of [Bus Stuffing Demos](https://atariage.com/forums/topic/258191-bus-stuffing-demos/):

1. RPG driver (newest driver)
2. Draconian driver (oldest driver)
3. all others 

and have updated Stella to detect and run them.

Still need to update the debugger tabs for BUS, clean up the code, then remember how to submit the changes. It's been so long since I last did anything with github that I just issued the following in Terminal to get the latest source:
`git clone https://github.com/stella-emu/stella.git`

<img width="640" alt="Screen Shot 2022-06-06 at 6 51 50 PM" src="https://user-images.githubusercontent.com/26531922/172267664-04171cc9-9504-40c1-a8ef-bd05fcf8c59b.png">

<img width="640" alt="Screen Shot 2022-06-06 at 6 52 36 PM" src="https://user-images.githubusercontent.com/26531922/172267680-56ca8f16-4275-4dd5-92ea-813607ffbbcb.png">

<img width="640" alt="Screen Shot 2022-06-06 at 6 49 44 PM" src="https://user-images.githubusercontent.com/26531922/172267697-47c25dac-3675-4ecc-abda-2a627f8d4cf7.png">

**Comment by thrust26** [comment]
> Thanks! Made some good headway this evening. So far I've identified 3 versions of the BUS driver in the demos that are linked to in the first post of [Bus Stuffing Demos](https://atariage.com/forums/topic/258191-bus-stuffing-demos/):
> 
> 1. RPG driver (newest driver)
> 2. Draconian driver (oldest driver)
> 3. all others
> 
> and have updated Stella to detect and run them.

> Still need to update the debugger tabs for BUS, clean up the code, then remember how to submit the changes. 

That sounds good. 👍 

> It's been so long since I last did anything with github that I just issued the following in Terminal to get the latest source: `git clone https://github.com/stella-emu/stella.git`

I don't work with GitHub for committing my changes too. Commandline or e.g. TortoiseGit are perfect.

**Comment by thrust26** [comment]
@SpiceWare Is bus stuffing really dead? Or are you still working on a solution?

**Comment by SpiceWare** [comment]
I'm not working on a solution, it's beyond my skill set. Chris was the one working on the BUS driver, with help from Fred. On Friday I was talking with Chris about Boom!, he mentioned: "I don't think we will get stable bus-stuffing any time soon, so I was glad to be able to reboot the game with CDFJ."

Both Chris and John have rebooted their BUS projects to use CDFJ. My most recent changes to Stella were to update CDFJ with the optional ability to override LDX # and/or LDY # for Fast Fetcher use. That change was for John's conversion of his BUS project, Elevator Action, to CDFJ.

Last week's conversation with Chris was the tipping point for me to decide that it didn't make sense to spend a lot of time on the BUS projects. I decided to update Stella as that will take significantly less time than updating all the BUS projects.

**Comment by sa666666** [comment]
@SpiceWare, we are planning on the 6.7 release for Saturday, June 11.  Do you think you can get this stuff included if I push the release to Monday, June 13 instead?

**Comment by thrust26** [comment]
@SpiceWare Agreed, if there are newer ROMs available,which work with your changes, then I wouldn't invest too much time.

**Comment by SpiceWare** [comment]
@sa666666 There's not much left if I'm not going to worry about those 2 old demos, so I'll be able to get the changes checked in by tomorrow night.

I've already shifted the BUS version numbers by 1. BUS0 is now used for the older demos shows an unsupported message in the debugger.

Is there another way to flag a ROM as unsupported so Stella shows a message when you attempt to load the ROM?

<img width="1440" alt="Screen Shot 2022-06-09 at 12 57 17 PM" src="https://user-images.githubusercontent.com/26531922/172913547-43bdc03f-d97d-46e6-af22-5f5ce6601fe9.png">

**Comment by thrust26** [comment]
> Is there another way to flag a ROM as unsupported so Stella shows a message when you attempt to load the ROM?

Maybe `setMessageCallback` would work for you.

**Comment by sa666666** [comment]
@SpiceWare, can you also post/send all the BUS ROMs you've tested?

**Comment by thrust26** [comment]
There are a number in the test folder. Which ones are you missing?

**Comment by sa666666** [comment]
Yes, committing them is fine too.

**Comment by SpiceWare** [comment]
Will commit the additional test ROMs with the exception of the 3 I'm not free to hand out.  Should I include my notes text file about the identified versions?

I'd contacted Chris and John this morning about my BUS update for Stella:

John sent me the BUS version of Elevator Action to test, but he doesn't want it released until after he's revealed the CDFJ reboot. After that he'll post the BUS version in the Bus Stuffing Demos topic.

Haven't heard back from Chris yet, though I did find a couple early versions of Boom! that he'd previously sent me via PM. I don't think he's publicly released those.

**Comment by thrust26** [comment]
@SpiceWare Do the ROMs you cannot post get autodetected by Stella? Else, maybe add the properties (in Stella.pro) too.

**Comment by SpiceWare** [comment]
I just checked Stella.pro, **"Cart.Type" "BUS"** is not in it so all the BUS ROMS have been autodetected by Stella.

**Comment by sa666666** [comment]
Yes, BUS ROMs are autodetected with the following code:
```
  // BUS ARM code has 2 occurrences of the string BUS
  // Note: all Harmony/Melody custom drivers also contain the value
  // 0x10adab1e (LOADABLE) if needed for future improvement
  static constexpr uInt8 bus[] = { 'B', 'U', 'S'};
  return searchForBytes(image, size, bus, 3, 2);
```

**Comment by sa666666** [comment]
@SpiceWare, I've decided to push the release until Monday, June 13, as I'd like to get your changes included in it.

**Comment by SpiceWare** [comment]
@sa666666 I believe I've finished my changes. I'd done the command

`git clone https://github.com/stella-emu/stella.git`

in Terminal, so can't submit my changes from my current workspace.  Tomorrow I'll get my GitHub fork up to date, get a fresh copy of the source from my fork, then move the changed files over so I can submit the changes.

**Comment by SpiceWare** [comment]
Since the release has been pushed to Monday I've looked at the original BUS demos some more.

During lunch I was able to make some changes that prevent Stella from immediately "crashing" the older demos into the debugger. Big part of that was due to the newer (supported) BUS drivers being 2K in size, while the older (unsupported) drivers are 3K. This difference caused the 4K banks of 6507 code to be misaligned by 1K for the older BUS demos. 

Additionally the new BUS routines support banks 0-6, and defaulted to bank 6, while the older BUS only has banks 0-5.

Hopefully I can get the the older demos working this weekend.  If not I'll check in what is working and leave the older BUS versions as "unsupported" for the time being.

**Comment by thrust26** [comment]
@SpiceWare Sounds good. But maybe you want to do a pull request before you start working on the older demos again. Then we could already review the existing code. 

Later you can either update the pull request (if we did not pull) or create a new one.

**Comment by SpiceWare** [comment]
I've submitted the changes. It's not 100% but is far better than what was there before.  For the oldest demos (identified as BUS0) I was able to get 1 of them working, while the others no longer "crash" into the debugger.

We have unexpected company in town from Mexico, so I won't have much additional time to spend on it this weekend.

<img width="1440" alt="Screen Shot 2022-06-11 at 6 01 38 PM" src="https://user-images.githubusercontent.com/26531922/173207600-6210b5f3-3148-49fb-a128-b8324d420000.png">

**Comment by sa666666** [comment]
@SpiceWare, thanks for working on this.

**Comment by SpiceWare** [comment]
You're welcome. Looks like a check failed though.  

https://ci.appveyor.com/project/sa666666/stella/builds/43831654

**Comment by sa666666** [comment]
Yes, that's fine.  It's because the newly added file wasn't added to Visual Studio.  I'm working on that now.

**Comment by thrust26** [comment]
@SpiceWare Thanks for your updates. I hope my Pull request instructions helped a bit.

I wonder why all ROMs are silent. Is there still something missing?

**Comment by SpiceWare** [comment]
Only BUS3 has working audio. I have some additional test ROMs but they name clashed so I didn't add them.

I'd deleted my Fork then recreated it before submitting my changes.  I just attempted to submit the BUS3 ROMs, I'd renamed them to have a **bus3_** prefix, but no luck as I'm in that weird state of:
`This branch is 2 commits ahead, 8 commits behind stella-emu:master.`

One of those that I'm ahead is my BUS changes, which are in stella-emu:master so I don't understand why that's considered "ahead". The other is the bus3_ ROMs.  

I tried doing a Fetch and Merge but there are conflicts that must be resolved via Command Line. I've not done that before, and don't have time to look into at the moment as I'm getting ready to head over to my folks'.

Anyway here's the additional BUS3 ROMs. Audio and speech both work.

[bus3_ROMs.zip](https://github.com/stella-emu/stella/files/8886196/bus3_ROMs.zip)

**Comment by sa666666** [comment]
I've added these ROMs.  @SpiceWare, can we close this one now, or will you continue to work on it after the upcoming release?

**Comment by thrust26** [comment]
> I'd deleted my Fork then recreated it before submitting my changes. I just attempted to submit the BUS3 ROMs, I'd renamed them to have a **bus3_** prefix, but no luck as I'm in that weird state of: `This branch is 2 commits ahead, 8 commits behind stella-emu:master.`

Which means, your local branch has 2 commits which are not synced to the master. And the master has 8 commit, which are not synced to you local branch.

> One of those that I'm ahead is my BUS changes, which are in stella-emu:master so I don't understand why that's considered "ahead". The other is the bus3_ ROMs.

Git is not tracking the individual changes, but the commits. 

> I tried doing a Fetch and Merge but there are conflicts that must be resolved via Command Line. I've not done that before, and don't have time to look into at the moment as I'm getting ready to head over to my folks'.

Yes, that's a bit more complicated. I am using TortoiseGit here, as I don't like to learn command line commands in 2022.

**Comment by sa666666** [comment]
Before doing a commit, always do a `git pull` to bring your local repo up to date with the remote one.  Then do the commit/push.  That will fix the error you're seeing.

**Comment by SpiceWare** [comment]
@sa666666 I'll look at it some more - I tracked down a later version of Stella's source that generates audio for **test1_music.bin**, one of the BUS2 demos. The audio is a bit distorted, but can still clearly make out the music. 

Based on this from **bus.h** for the older 3K BUS driver, BUS0 did not have audio routines implemented so I should remove the Music registers in the debugger for BUS0.
```
UNUSED0     DS 1    ; $14    AMPLITUDE?
UNUSED1     DS 1    ; $15    WAVEFORM0?
UNUSED2     DS 1    ; $16    WAVEFORM1?
UNUSED3     DS 1    ; $17    WAVEFORM2?
UNUSED4     DS 1    ; $18    NOTE0?
UNUSED5     DS 1    ; $19    NOTE1?
UNUSED6     DS 1    ; $1A    NOTE2?
UNUSED7     DS 1    ; $1B
```

**Comment by sa666666** [comment]
@SpiceWare, pinging to see if you're still working on this, or if we can close it.

**Comment by SpiceWare** [comment]
I've not looked at it in years, go ahead and close it.
